File: Python/A_Project/A_Tic_Tac_Toe/GameStateHandler/ConcreteStates/x_turn_state.py
from CentralEnum.symbol import Symbol
from GameStateHandler.game_state import GameState


class XTurnState(GameState):
    def next(self, context, player, has_won: bool):
        if has_won:
            from GameStateHandler.ConcreteStates.x_won_state import XWonState
            context.set_state(XWonState())
        else:
            from GameStateHandler.ConcreteStates.o_turn_state import OTurnState

            context.set_state(OTurnState())

    def is_game_over(self) -> bool:
        return False

    def get_result_message(self) -> str:
        return "Game is still in progress."

File: Ruby/A_Project/A_Tic_Tac_Toe/GameStateHandler/ConcreteStates/x_turn_state.rb
require_relative '../game_state'

module GameStateHandler
  module ConcreteStates
    class XTurnState
      include GameState

      def next(context, player, has_won)
        # After a player has won we transition to the appropriate winning state.
        # context is needed here to allow clean state transition.
        if has_won
          require_relative './x_won_state'
          require_relative './o_won_state'
          context.state = (player.symbol == CentralEnum::Symbol::X ? XWonState.new : OWonState.new)
        else
          require_relative './o_turn_state'
          context.state = OTurnState.new
        end
      end

      def game_over?
        false
      end

      def result_message
        "Game is still in progress."
      end
    end
  end
end
